_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1144 @@
+
+## Update the User Model
+
+ When integrating with Okta, we will need to track whether each user is active. To track this, we'll add an `active` field to the Prisma database schema, specifically the user table. Here we will add the field attribute `active` and set that to be a Boolean to inform us whether or not the user is active. Although not required by the SCIM protocol as a core attribute, we will need this later when connecting with Okta.
+
+Speaking of attributes, the SCIM core user schema requires only three attributes (all other attributes are optional):
+
+- `id`, service provider-defined identifier
+- `userName`, a unique identifier for the user 
+- `meta`, read-only metadata maintained by the service provider
+
+These are described in section 8.1 of the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7643#section-8.1). 
+
+Note: We'll be using a basic autoincrement for the users' id to lessen the complexity of this project. However, for production, we recommend using a unique id generator such as [xid](https://www.npmjs.com/package/xid-js). 
+
+Update the user model in `schema.prisma`: 
+
+```
+model User {
+  id       Int    @id @default(autoincrement())
+  orgId    Int?
+  externalId String?
+  email    String
+  password String?
+  name     String
+  Todo     Todo[]
+  org      Org?    @relation(fields: [orgId],  references: [id])
+  active   Boolean?
+ @@unique([orgId, externalId])
+}
+```
+
+Now that we have the user table with the necessary fields, we'll need to edit the `seed_script.ts` file. In this workshop, I'll show you how to create multiple separate organizations and users as needed. Provisioning multiple separate orgs accommodates your customers who may be using different identity providers with different SCIM clients. By giving each organization its own endpoint and API token, we ensure that no org can accidentally update or overwrite another org's users.
+
+Let's update the users in `seed.ts`. We'll also need to hardcode `externalId` and `active` for both users. Note that `externalId`  is a unique identifier issued by the provisioning client and must be stored in the SCIM server. The `externalId` is also considered stable because a user's name and email address can change. Along with the `active` attribute, `externalId` is not a SCIM protocol core attribute, but we will need it when we connect with Okta. In Okta, the  [Okta SCIM Docs](https://developer.okta.com/docs/guides/scim-provisioning-integration-prepare/main/#basic-user-schema) tell us that this unique identifier is the user's global id (GUID). Ref:
+
+We'll also set an `apikey` to an arbitrary string, because it will be used when we implement token auth to test each route as we build it.
```

Comment:
```suggestion
We'll also set an `apikey` to an random string, because it will be used when we implement token auth to test each route as we build it.
```